src/pybind/mgr/volumes/volume.py

```diff
@@ -0,0 +1,151 @@
+import json
+import errno
+import logging
+import orchestrator
+
+log = logging.getLogger(__name__)
+
+class VolumeClient(object):
+    def __init__(self, mgr):
+        """
+        """
+        self.mgr = mgr
+
+    def gen_pool_names(self, volname):
+        """
+        """
+        return "cephfs.{}.meta".format(volname), "cephfs.{}.data".format(volname)
+
+    def get_fs(self, fs_name):
+        fs_map = self.mgr.get('fs_map')
+        for fs in fs_map['filesystems']:
+            if fs['mdsmap']['fs_name'] == fs_name:
+                return fs
+        return None
+
+    def get_mds_names(self, fs_name):
+        fs = self.get_fs(fs_name)
+        if fs is None:
+            return []
+        return [mds['name'] for mds in fs['mdsmap']['info'].values()]
+
+    def volume_exists(self, volname):
+        return self.get_fs(volname) is not None
+
+    def create_pool(self, pool_name, pg_num, pg_num_min=None, pg_autoscale_factor=None):
+        # create the given pool
+        command = {'prefix': 'osd pool create', 'pool': pool_name, 'pg_num': pg_num}
+        if pg_num_min is not None:
+            command['pg_num_min'] = pg_num_min
+        r, outb, outs = self.mgr.mon_command(command)
+        if r != 0:
+            return r, outb, outs
+
+        # set pg autoscale if needed
+        if pg_autoscale_factor is not None:
+            command = {'prefix': 'osd pool set', 'pool': pool_name, 'var': 'pg_autoscale_bias',
+                       'val': str(pg_autoscale_factor)}
+            r, outb, outs = self.mgr.mon_command(command)
+        return r, outb, outs
+
+    def remove_pool(self, pool_name):
+        command = {'prefix': 'osd pool rm', 'pool': pool_name, 'pool2': pool_name,
+                   'yes_i_really_really_mean_it': True}
+        return self.mgr.mon_command(command)
+
+    def create_filesystem(self, fs_name, metadata_pool, data_pool):
+        command = {'prefix': 'fs new', 'fs_name': fs_name, 'metadata': metadata_pool,
+                   'data': data_pool}
+        return self.mgr.mon_command(command)
+
+    def remove_filesystem(self, fs_name):
+        command = {'prefix': 'fs rm', 'fs_name': fs_name, 'yes_i_really_mean_it': True}
+        return self.mgr.mon_command(command)
+
+    def create_mds(self, fs_name):
+        spec = orchestrator.StatelessServiceSpec()
+        spec.name = fs_name
+        try:
+            completion = self.mgr.add_stateless_service("mds", spec)
+            self.mgr._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+        except (ImportError, orchestrator.OrchestratorError):
+            return 0, "", "Volume created successfully (no MDS daemons created)"
+        except Exception as e:
+            # Don't let detailed orchestrator exceptions (python backtraces)
+            # bubble out to the user
```

Comment:
Can we add this log message as before,
self.log.exception("Failed to create MDS daemons")